homeassistant/components/goodwe/switch.py

```diff
@@ -0,0 +1,131 @@
+from __future__ import annotations
+
+from dataclasses import dataclass
+import logging
+from typing import Any
+
+from goodwe import Inverter, InverterError
+
+from homeassistant.components.switch import (
+    SwitchDeviceClass,
+    SwitchEntity,
+    SwitchEntityDescription,
+)
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import EntityCategory
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.device_registry import DeviceInfo
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+
+from .const import DOMAIN, KEY_DEVICE_INFO, KEY_INVERTER
+
+_LOGGER = logging.getLogger(__name__)
+
+
+@dataclass(frozen=True)
+class GoodweSwitchEntityDescriptionBase:
+    """Required values when describing Goodwe switch settings."""
+
+    icon_on: str
+    icon_off: str
+
+
+@dataclass(frozen=True)
+class GoodweNumberEntityDescription(
+    SwitchEntityDescription, GoodweSwitchEntityDescriptionBase
+):
+    """Class describing Goodwe switch entities."""
+
+
+SWITCHES = (
+    # Export limit Enabled
+    GoodweNumberEntityDescription(
+        key="grid_export",
+        translation_key="grid_export",
+        icon_on="mdi:transmission-tower-off",
+        icon_off="mdi:transmission-tower-export",
+        entity_category=EntityCategory.CONFIG,
+        device_class=SwitchDeviceClass.SWITCH,
+    ),
+)
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    config_entry: ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up the inverter switch entities from a config entry."""
+    inverter = hass.data[DOMAIN][config_entry.entry_id][KEY_INVERTER]
+    device_info = hass.data[DOMAIN][config_entry.entry_id][KEY_DEVICE_INFO]
+
+    entities = []
+
+    for description in SWITCHES:
+        try:
+            current_value = await inverter.read_setting(description.key)
+        except (InverterError, ValueError):
+            # Inverter model does not support this setting
+            _LOGGER.debug("Could not read inverter setting %s", description.key)
+            continue
+
+        entities.append(
+            InverterSwitchEntity(device_info, description, inverter, current_value == 1)
+        )
+
+    async_add_entities(entities)
+
+
+class InverterSwitchEntity(SwitchEntity):
+    """Inverter switch setting entity."""
+
+    entity_description: GoodweNumberEntityDescription
+
+    _attr_should_poll = False
+    _attr_has_entity_name = True
+
+    def __init__(
+        self,
+        device_info: DeviceInfo,
+        description: GoodweNumberEntityDescription,
+        inverter: Inverter,
+        current_state: bool,
+    ) -> None:
+        """Initialize the inverter switch setting entity."""
+        self.entity_description = description
+        self._attr_unique_id = f"{DOMAIN}-{description.key}-{inverter.serial_number}"
+        self._attr_device_info = device_info
+        self._attr_is_on = current_state
```

Comment:
Schouldn't we poll for the state of the switch every x minutes?
Now if someone changes the value in the goodwe app/SEMS app, it will never be updated in HomeAssistant untill you restart the integration....

Reply:
I used the same logic as for grid_export_limit. these can only be changed via the PV master app (and maybe another app for different inverters), but these are not accessible in the SEMS portal (as of my knowledge).

I will implement the async_update

Reply:
@starkillerOG Actually, I probably wouldn't poll (just) the switch state.
There are already other "settings" (not sensors) managed by the HA (like DoD, export limit) and those are so far NOT polled from inverter.
The approach was simple - one should manage the inverter either by HA or by its mobile app.
If we add this polling/synchronization just here, it would be confusing why something is and something is not synchronized.
So I suggest to remove the polling from here for now and if we really intend to implement it, then we should do it for all settings, not just switch ...

Reply:
The proper thing in my opinion would then be to create a dataupdatecoordinator and start polling those settings at a low poll interval (once every hour or so) to at least keep it in sync.